libdrgn/handler.c

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "cleanup.h"
+#include "drgn.h"
+#include "handler.h"
+#include "hash_table.h"
+#include "util.h"
+
+void drgn_handler_list_deinit(struct drgn_handler_list *list)
+{
+	struct drgn_handler *handler = list->registered;
+	while (handler) {
+		struct drgn_handler *next = handler->next_registered;
+		if (handler->free) {
+			free((char *)handler->name);
+			free(handler);
+		}
+		handler = next;
+	}
+}
+
+struct drgn_error *drgn_handler_list_register(struct drgn_handler_list *list,
+					      struct drgn_handler *handler,
+					      size_t enable_idx,
+					      const char *what)
+{
+	struct drgn_handler **registeredp = &list->registered;
+	while (*registeredp) {
+		if (strcmp(handler->name, (*registeredp)->name) == 0) {
+			return drgn_error_format(DRGN_ERROR_INVALID_ARGUMENT,
+						 "duplicate %s name '%s'",
+						 what, handler->name);
+		}
+		registeredp = &(*registeredp)->next_registered;
+	}
+	handler->next_registered = NULL;
+	*registeredp = handler;
+	if (enable_idx != DRGN_HANDLER_REGISTER_DONT_ENABLE) {
+		struct drgn_handler **enabledp = &list->enabled;
+		for (size_t i = 0; i < enable_idx && *enabledp; i++)
+			enabledp = &(*enabledp)->next_enabled;
+		handler->next_enabled = *enabledp;
+		*enabledp = handler;
+	}
+	return NULL;
+}
+
+struct drgn_error *drgn_handler_list_registered(struct drgn_handler_list *list,
+						const char ***names_ret,
+						size_t *count_ret)
+{
+	size_t n = 0;
+	for (struct drgn_handler *handler = list->registered; handler;
+	     handler = handler->next_registered)
+		n++;
+	const char **names = malloc_array(n, sizeof(names[0]));
+	if (!names)
+		return &drgn_enomem;
+	size_t i = 0;
+	for (struct drgn_handler *handler = list->registered; handler;
+	     handler = handler->next_registered)
+		names[i++] = handler->name;
+	*names_ret = names;
+	*count_ret = n;
+	return NULL;
+}
```

Comment:
Logic here could be shared with the `_enabled` version below, if you don't mind the preprocessor magic:
```
#define HANDLER_LIST_FN(ATTRIB) \
	struct drgn_error *drgn_handler_list_ ## ATTRIB(struct drgn_handler_list *list, \
							const char ***names_ret, \
							size_t *count_ret) \
	{ \
		size_t n = 0; \
		for (struct drgn_handler *handler = list->ATTRIB; handler; \
		     handler = handler->next_ ## ATTRIB) \
			n++; \
		const char **names = malloc_array(n, sizeof(names[0])); \
		if (!names) \
			return &drgn_enomem; \
		size_t i = 0; \
		for (struct drgn_handler *handler = list->ATTRIB; handler; \
		     handler = handler->next_ ## ATTRIB) \
			names[i++] = handler->name; \
		*names_ret = names; \
		*count_ret = n; \
		return NULL; \
	}
HANDLER_LIST_FN(registered);
HANDLER_LIST_FN(enabled);
```